File: c++/0077.cpp
// Link: https://leetcode.com/problems/combinations/description/
// Runtime: 403 ms
// Memory: 164.87 MB

class Solution 
{
public:
    std::vector<std::vector<int>> combine(int n, int k) 
    {
        for (short i = 1; i <= n; i++) range.push_back(i);
        
        addCombination({}, 0, 0, k);
        
        return combinations;
    }

private:
    std::vector<short> range;
    std::vector<std::vector<int>> combinations;

    void addCombination(std::vector<int> combination, int value, int depth, int maxDepth)
    {
        if (depth == maxDepth) combinations.push_back(combination);
        else
            for (int i = value; i < static_cast<int>(range.size()); i++)
            {
                combination.push_back(range[i]);
                addCombination(combination, range[i], depth + 1, maxDepth);
                combination.pop_back();
            }
    }
};

File: python/0077.py
# Link: https://leetcode.com/problems/combinations/description/
# Runtime: 362 ms
# Memory: 18.2 MB


class Solution:
    range_of_elements = []
    combinations = []

    def combine(self, n: int, k: int) -> List[List[int]]:
        self.range_of_elements = [element for element in range(1, n + 1)]
        self.combinations = []

        self.__add_combination([], 0, 0, k)

        return self.combinations

    def __add_combination(
        self,
        combination: List[int],
        value_for_next_element: int,
        depth: int,
        max_depth: int,
    ) -> None:
        if depth == max_depth:
            self.combinations.append(combination[:])
        else:
            for i in range(value_for_next_element, len(self.range_of_elements)):
                combination.append(self.range_of_elements[i])
                self.__add_combination(
                    combination, self.range_of_elements[i], depth + 1, max_depth
                )
                combination.pop()
